### PR TITLE
Create base reconciler for deletion

### PR DESCRIPTION
**Pull request by conradoverta**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @Override
+  public void resync() {
+    String queryString =
+        String.format(
+            "select id from %s where deleted=:deleted", ExperimentRunEntity.class.getSimpleName());
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      Query deletedQuery = session.createQuery(queryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]:
```suggestion
      Query<String> deletedQuery = session.createQuery(queryString, String.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  protected void reconcile(Set<String> ids) {
+    LOGGER.debug("Reconciling experiment runs " + ids.toString());
+
+    deleteRoleBindings(ids);
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String experimentRunQueryString =
+          String.format("from %s where id=:ids", ExperimentRunEntity.class.getSimpleName());
+
+      Query experimentRunDeleteQuery = session.createQuery(experimentRunQueryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]:
```suggestion
      Query<ExperimentRunEntity> experimentRunDeleteQuery = session.createQuery(experimentRunQueryString, ExperimentRunEntity.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      LOGGER.error("reconcile: ", ex);
+    }
+  }
+
+  private void deleteRoleBindings(Set<String> ids) {
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String deleteExperimentRunQueryString =
+          String.format("FROM %s WHERE id IN (:ids)", ExperimentRunEntity.class.getSimpleName());
+
+      Query experimentRunDeleteQuery = session.createQuery(deleteExperimentRunQueryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]:
```suggestion
      Query<ExperimentRunEntity> experimentRunDeleteQuery = session.createQuery(deleteExperimentRunQueryString, ExperimentRunEntity.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  @Override
+  public void resync() {
+    String queryString =
+        String.format(
+            "select id from %s where deleted=:deleted", ExperimentEntity.class.getSimpleName());
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      Query deletedQuery = session.createQuery(queryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java]:
```suggestion
      Query<String> deletedQuery = session.createQuery(queryString, String.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  protected void reconcile(Set<String> ids) {
+    LOGGER.debug("Reconciling experiments " + ids.toString());
+
+    deleteRoleBindings(ids);
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String experimentQueryString =
+          String.format("from %s where id=:ids", ExperimentEntity.class.getSimpleName());
+
+      Query experimentDeleteQuery = session.createQuery(experimentQueryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java]:
```suggestion
      Query<ExperimentEntity> experimentDeleteQuery = session.createQuery(experimentQueryString, ExperimentEntity.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      LOGGER.error("reconcile: ", ex);
+    }
+  }
+
+  private void deleteRoleBindings(Set<String> ids) {
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String deleteExperimentQueryString =
+          String.format("FROM %s WHERE id IN (:ids)", ExperimentEntity.class.getSimpleName());
+
+      Query experimentDeleteQuery = session.createQuery(deleteExperimentQueryString);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java]:
```suggestion
      Query<ExperimentEntity> experimentDeleteQuery = session.createQuery(deleteExperimentQueryString, ExperimentEntity.class);
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+  @Override
+  protected void reconcile(Set<String> ids) {
+    LOGGER.debug("Reconciling experiment runs " + ids.toString());
+
+    deleteRoleBindings(ids);
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String experimentRunQueryString =
+          String.format("from %s where id=:ids", ExperimentRunEntity.class.getSimpleName());
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]:
```suggestion
          String.format("from %s where id in (:ids)", ExperimentRunEntity.class.getSimpleName());
```

Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]:
```suggestion
          String.format("from %s where id in (:ids)", ExperimentRunEntity.class.getSimpleName());
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Override
+  protected void reconcile(Set<String> ids) {
+    LOGGER.debug("Reconciling experiments " + ids.toString());
+
+    deleteRoleBindings(ids);
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String experimentQueryString =
+          String.format("from %s where id=:ids", ExperimentEntity.class.getSimpleName());
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperiments.java]:
```suggestion
          String.format("from %s where id in (:ids)", ExperimentEntity.class.getSimpleName());
```

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteProjects.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  @Override
+  protected void reconcile(Set<String> ids) {
+    LOGGER.debug("Reconciling projects " + ids.toString());
+
+    roleService.deleteEntityResourcesWithServiceUser(
+        new ArrayList<>(ids), ModelDBResourceEnum.ModelDBServiceResourceTypes.PROJECT);
+
+    try (Session session = modelDBHibernateUtil.getSessionFactory().openSession()) {
+      String projectsQueryString =
+          String.format("from %s where id=:ids", ProjectEntity.class.getSimpleName());
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteProjects.java]:
```suggestion
          String.format("from %s where id in (:ids)", ProjectEntity.class.getSimpleName());
```

**Review by evgeniilezhnin — COMMENTED** [review]
Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?

File: backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java (hunk 1 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/reconcilers/SoftDeleteExperimentRuns.java]: Didn't know about this! Pretty cool.

**Comment by conradoverta** [comment]
> Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?

Not sure what you mean. Could you expand?

**Comment by evgeniilezhnin** [comment]
> > Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?
> 
> Not sure what you mean. Could you expand?

I mean queries for different entity types look alike. Should we reuse it?

**Comment by conradoverta** [comment]
> > > Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?
> > 
> > 
> > Not sure what you mean. Could you expand?
> 
> I mean queries for different entity types look alike. Should we reuse it?

I still don't follow what you mean. You can use whatever queries are necessary to sync the state and to process one set of changes.

**Review by evgeniilezhnin — APPROVED** [review]

**Comment by evgeniilezhnin** [comment]
> > > > Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?
> > > 
> > > 
> > > Not sure what you mean. Could you expand?
> > 
> > 
> > I mean queries for different entity types look alike. Should we reuse it?
> 
> I still don't follow what you mean. You can use whatever queries are necessary to sync the state and to process one set of changes.

I mean when there is a change needed in one of the queries, there is also the same change needed in another. So maybe there is a way to do these actions using some template. Anyway it's not urgent, so can be done later if necessary.

**Comment by conradoverta** [comment]
> > > > > Is it possible to reuse Reconciler implementation queries code for example using Criteria Queries?
> > > > 
> > > > 
> > > > Not sure what you mean. Could you expand?
> > > 
> > > 
> > > I mean queries for different entity types look alike. Should we reuse it?
> > 
> > 
> > I still don't follow what you mean. You can use whatever queries are necessary to sync the state and to process one set of changes.
> 
> I mean when there is a change needed in one of the queries, there is also the same change needed in another. So maybe there is a way to do these actions using some template. Anyway it's not urgent, so can be done later if necessary.

Ah, yes. I tried doing a DBReconciler but the differences were significant enough to make that extra abstraction more complex than simpler =)

**Pull request merged**